x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/LookupJoinTypesIT.java

```diff
@@ -0,0 +1,463 @@
+ * </pre>
+ * And assert that the result exists and is equal to "value".
+ */
+@ClusterScope(scope = SUITE, numClientNodes = 1, numDataNodes = 1)
+public class LookupJoinTypesIT extends ESIntegTestCase {
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(EsqlPlugin.class, MapperExtrasPlugin.class);
+    }
+
+    private static final Map<String, TestConfigs> testConfigurations = new HashMap<>();
+    static {
+        // Initialize the test configurations for string tests
+        {
+            TestConfigs configs = testConfigurations.computeIfAbsent("strings", TestConfigs::new);
+            configs.addPasses(KEYWORD, KEYWORD);
+            configs.addPasses(TEXT, KEYWORD);
+            configs.addFailsText(KEYWORD);
+            configs.addFailsText(TEXT);
+        }
+
+        // Test integer types
+        var integerTypes = List.of(BYTE, SHORT, INTEGER);
+        {
+            TestConfigs configs = testConfigurations.computeIfAbsent("integers", TestConfigs::new);
+            for (DataType mainType : integerTypes) {
+                for (DataType lookupType : integerTypes) {
+                    configs.addPasses(mainType, lookupType);
+                }
+                // Long is currently treated differently in the validation, but we could consider changing that
+                configs.addFails(mainType, LONG);
+                configs.addFails(LONG, mainType);
+            }
+        }
+
+        // Test float and double
+        var floatTypes = List.of(HALF_FLOAT, FLOAT, DOUBLE);
+        {
+            TestConfigs configs = testConfigurations.computeIfAbsent("floats", TestConfigs::new);
+            for (DataType mainType : floatTypes) {
+                for (DataType lookupType : floatTypes) {
+                    configs.addPasses(mainType, lookupType);
+                }
+            }
+        }
+
+        // Tests for mixed-numerical types
+        {
+            TestConfigs configs = testConfigurations.computeIfAbsent("mixed-numerical", TestConfigs::new);
+            for (DataType mainType : integerTypes) {
+                for (DataType lookupType : floatTypes) {
+                    // TODO: We should probably allow this, but we need to change the validation code in Join.java
+                    configs.addFails(mainType, lookupType);
+                    configs.addFails(lookupType, mainType);
+                }
+            }
+        }
+
+        // Tests for all types where left and right are the same type
+        DataType[] all = { BOOLEAN, LONG, INTEGER, DOUBLE, SHORT, BYTE, FLOAT, HALF_FLOAT, DATETIME, IP, KEYWORD };
+        {
+            Collection<TestConfigs> existing = testConfigurations.values();
+            TestConfigs configs = testConfigurations.computeIfAbsent("same", TestConfigs::new);
+            for (DataType type : all) {
+                if (existingIndex(existing, type, type)) {
```

Comment:
Supernit: test against false and skip the `continue` for slightly better legibility.
Here and below.

Reply:
Done